### ISSUE TITLE
Safe list destructuring

### ISSUE THREAD

**Issue by Reprevise** [dart-lang/language#3430]
Let's say I have a variable called `list` that stores a, you guessed it, list of items.

```dart
final list = ['a', 'b'];
```

For this example, we're always going to assume that list at least has a length of 1. When we need to safely access the second value in the list though, it's kind of annoying.

```dart
final a = list.first;
final b = list.elementAtOrNull(1);
```

Here are some proposed syntax additions that can help make it easier to safely access list items:

```dart
final [a, b?] = list; // b would be null if the list doesn't have enough items
```

For the syntax below, you can specify default values, but they must be a subtype of the inferred list type. So `list` is inferred as `List<String>` so you can do default values of type `String?`.

Using `else`:
```dart
final [a, b else null] = list;
```

Using `??`:
```dart
final [a, b ?? null] = list; // looks odd but could work, reads as if not present, use null
```

Default value syntax used in constructors:
```dart
final [a, b = null] = list;
```

**Comment by lrhn** [comment]
The default-value syntax or `else` syntax are the most viable, the `??` one looks like it should act on `null`, not absence.

Also makes sense for map patterns, as a way to match an entry if it's there, but also continue is it's not.

**Comment by Reprevise** [comment]
> the ?? one looks like it should act on null, not absence.

Yeah I'm just throwing out ideas, it didn't look great to me either. I also had a `b : null` syntax idea but I thought it'd be too bad of an idea to post here lol.

> Also makes sense for map patterns, as a way to match an entry if it's there, but also continue is it's not.

Agreed, I could file a separate issue for that, or make this issue more generic. That could also apply if it wasn't the type you were expecting, instead of throwing an exception, but that behavior seems unclear, though it could make programs safer.

> Not sure how it generalizes to refutable patterns. As written it makes sense for a binding pattern. It must be at the top-level list element pattern, since it's what triggers of there is no matched value, and we have to make that choice before trying any pattern that has to use that value.

I assume this is in response to `final [a, b?] = list;` which I actually tried to write (it's what inspired this issue), and to me seems natural to read. It's also the shortest syntax I proposed. The `?` is synonymous with nullable so I thought it just made sense.

**Comment by SupposedlySam** [comment]
Any movement on this?

## My Use Case
I'm writing some extension methods for cloud_firestore's `Filter` class that has a named constructor taking up to 30 arguments. I'd love to create an extension getter so I can have something like:

```dart
final myListFilters = [...some actual filters here];

myListOfFilters.asAndGroup;
```

In the `asAndGroup` method I'd want something like this:
```dart

    final [
      l1,
      l2,
      l3?,
      l4?,
     ... all the way to 30
    ] = this;

    return Filter.and(
      l1, // required `Filter`
      l2, // required `Filter`
      l3, // optional `Filter?`
      l4, // optional `Filter?`
      ... all the way to 30
    );
```
